swarm/benches/connection_handler.rs

```diff
@@ -0,0 +1,357 @@
+            .authenticate(libp2p_plaintext::Config::new(&swarm_a_keypair))
+            .multiplex(libp2p_yamux::Config::default())
+            .boxed(),
+        beh,
+        swarm_a_keypair.public().to_peer_id(),
+        libp2p_swarm::Config::with_tokio_executor().with_idle_connection_timeout(Duration::MAX),
+    )
+}
+
+#[derive(Default)]
+struct PollerBehaviour {
+    spam_count: usize,
+    protocols: &'static [StreamProtocol],
+    restarting: bool,
+    finished: bool,
+    other_peer: Option<PeerId>,
+}
+
+#[derive(Debug)]
+struct FinishedSpamming;
+
+impl NetworkBehaviour for PollerBehaviour {
+    type ConnectionHandler = PollerHandler;
+    type ToSwarm = FinishedSpamming;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: libp2p_swarm::ConnectionId,
+        _peer: libp2p_identity::PeerId,
+        _local_addr: &libp2p_core::Multiaddr,
+        _remote_addr: &libp2p_core::Multiaddr,
+    ) -> Result<libp2p_swarm::THandler<Self>, libp2p_swarm::ConnectionDenied> {
+        Ok(PollerHandler {
+            spam_count: 0,
+            protocols: self.protocols,
+        })
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _connection_id: libp2p_swarm::ConnectionId,
+        _peer: libp2p_identity::PeerId,
+        _addr: &libp2p_core::Multiaddr,
+        _role_override: libp2p_core::Endpoint,
+    ) -> Result<libp2p_swarm::THandler<Self>, libp2p_swarm::ConnectionDenied> {
+        Ok(PollerHandler {
+            spam_count: self.spam_count,
+            protocols: self.protocols,
+        })
+    }
+
+    fn on_swarm_event(&mut self, _: libp2p_swarm::FromSwarm) {}
+
+    fn on_connection_handler_event(
+        &mut self,
+        _peer_id: libp2p_identity::PeerId,
+        _connection_id: libp2p_swarm::ConnectionId,
+        _event: libp2p_swarm::THandlerOutEvent<Self>,
+    ) {
+        self.finished = true;
+    }
+
+    fn poll(
+        &mut self,
+        _: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<libp2p_swarm::ToSwarm<Self::ToSwarm, libp2p_swarm::THandlerInEvent<Self>>>
+    {
+        if self.finished && !self.restarting {
+            self.restarting = true;
+            std::task::Poll::Ready(libp2p_swarm::ToSwarm::GenerateEvent(FinishedSpamming))
+        } else if !self.finished && self.restarting {
+            self.restarting = false;
+            std::task::Poll::Ready(libp2p_swarm::ToSwarm::NotifyHandler {
+                peer_id: self.other_peer.unwrap(),
+                handler: libp2p_swarm::NotifyHandler::Any,
+                event: Restart(self.spam_count),
+            })
+        } else {
+            std::task::Poll::Pending
+        }
+    }
+}
+
+impl BigBehaviour for PollerBehaviour {
+    fn behaviours(&mut self) -> &mut [PollerBehaviour] {
+        std::slice::from_mut(self)
+    }
+}
+
+#[derive(Default)]
+struct PollerHandler {
+    spam_count: usize,
+    protocols: &'static [StreamProtocol],
```

Comment:
I'd prefer if we do something more production-like so this should be a `Vec` (which probably will make the results look even better).

Reply:
If this is a vec, I need to clone it in the connection handler to return protocols from the `fn listen_protocols()`, do we want to benchmark that?

Reply:
Yeah I think we should because that is what people do in production. Also, if we do it for both, the baseline and here the results should still be meaningful?